File: mobile/lib/services/routine_api.dart
import '../models/routine.dart';
import 'api_service.dart';

class RoutineApi {
  final ApiService _apiService;

  static RoutineApi get instance => routineApi;

  RoutineApi(this._apiService);

  // ==================== Routine CRUD ====================

  Future<List<Routine>> getAllRoutines() async {
    final response = await _apiService.get('/routines');
    final List<dynamic> data = response.data as List<dynamic>;
    return data.map((json) => Routine.fromJson(json as Map<String, dynamic>)).toList();
  }

  Future<RoutineDetail> getRoutineDetail({required int routineId}) async {
    final response = await _apiService.get('/routines/$routineId');
    return RoutineDetail.fromJson(response.data as Map<String, dynamic>);
  }

  Future<Routine> createRoutine({required String name}) async {
    final response = await _apiService.post(
      '/routines',
      data: {'name': name},
    );
    return Routine.fromJson(response.data as Map<String, dynamic>);
  }

  Future<Routine> updateRoutineName({
    required int routineId,
    required String name,
  }) async {
    final response = await _apiService.put(
      '/routines/$routineId/name',
      data: {'name': name},
    );
    return Routine.fromJson(response.data as Map<String, dynamic>);
  }

  Future<void> deleteRoutine({required int routineId}) async {
    await _apiService.delete('/routines/$routineId');
  }

  // ==================== Step CRUD ====================

  Future<RoutineStep> createStep({
    required int routineId,
    required String text,
    String? notes,
    int? position,
  }) async {
    final Map<String, dynamic> data = {'text': text};
    if (notes != null) data['notes'] = notes;
    if (position != null) data['position'] = position;

    final response = await _apiService.post(
      '/routines/$routineId/steps',
      data: data,
    );
    return RoutineStep.fromJson(response.data as Map<String, dynamic>);
  }

  Future<RoutineStep> updateStepText({
    required int routineId,
    required int stepId,
    required String text,
  }) async {
    final response = await _apiService.put(
      '/routines/$routineId/steps/$stepId/text',
      data: {'text': text},
    );
    return RoutineStep.fromJson(response.data as Map<String, dynamic>);
  }

  Future<RoutineStep> updateStepNotes({
    required int routineId,
    required int stepId,
    String? notes,
  }) async {
    final response = await _apiService.put(
      '/routines/$routineId/steps/$stepId/notes',
      data: {'notes': notes},
    );
    return RoutineStep.fromJson(response.data as Map<String, dynamic>);
  }

  Future<RoutineStep> updateStepPosition({
    required int routineId,
    required int stepId,
    required int position,
  }) async {
    final response = await _apiService.put(
      '/routines/$routineId/steps/$stepId/position',
      data: {'position': position},
    );
    return RoutineStep.fromJson(response.data as Map<String, dynamic>);
  }

  Future<void> deleteStep({
    required int routineId,
    required int stepId,
  }) async {
    await _apiService.delete('/routines/$routineId/steps/$stepId');
  }

  // ==================== Schedules ====================

  Future<List<RoutineSchedule>> setSchedules({
    required int routineId,
    required List<ScheduleEntry> schedules,
  }) async {
    final response = await _apiService.put(
      '/routines/$routineId/schedules',
      data: {'schedules': schedules.map((s) => s.toJson()).toList()},
    );
    final List<dynamic> data = response.data as List<dynamic>;
    return data.map((json) => RoutineSchedule.fromJson(json as Map<String, dynamic>)).toList();
  }

  // ==================== Prompts ====================

  Future<List<PendingRoutinePrompt>> getPendingPrompts() async {
    final response = await _apiService.get('/routines/prompts/pending');
    final List<dynamic> data = response.data as List<dynamic>;
    return data.map((json) => PendingRoutinePrompt.fromJson(json as Map<String, dynamic>)).toList();
  }

  Future<void> dismissPrompt({required int routineId}) async {
    await _apiService.post('/routines/prompts/$routineId/dismiss');
  }

  Future<RoutineCompletion> quickCompleteRoutine({required int routineId, List<int>? completedStepIds}) async {
    final Map<String, dynamic>? data = completedStepIds != null ? {'completedStepIds': completedStepIds} : null;
    final response = await _apiService.post('/routines/$routineId/quick-complete', data: data);
    return RoutineCompletion.fromJson(response.data as Map<String, dynamic>);
  }

  // ==================== Execution ====================

  Future<RoutineCompletion> startRoutine({required int routineId}) async {
    final response = await _apiService.post('/routines/$routineId/start');
    return RoutineCompletion.fromJson(response.data as Map<String, dynamic>);
  }

  Future<RoutineCompletion?> getActiveExecution({required int routineId}) async {
    final response = await _apiService.get('/routines/$routineId/active');
    if (response.statusCode == 204) return null;
    return RoutineCompletion.fromJson(response.data as Map<String, dynamic>);
  }

  Future<RoutineStepCompletion> completeStep({
    required int completionId,
    required int stepId,
    required String action,  // 'complete' or 'skip'
    String? notes,
  }) async {
    final Map<String, dynamic> data = {'action': action};
    if (notes != null) data['notes'] = notes;

    final response = await _apiService.post(
      '/routines/executions/$completionId/steps/$stepId',
      data: data,
    );
    return RoutineStepCompletion.fromJson(response.data as Map<String, dynamic>);
  }

  Future<RoutineCompletion> finishExecution({required int completionId}) async {
    final response = await _apiService.post('/routines/executions/$completionId/finish');
    return RoutineCompletion.fromJson(response.data as Map<String, dynamic>);
  }

  Future<RoutineCompletion> abandonExecution({required int completionId}) async {
    final response = await _apiService.post('/routines/executions/$completionId/abandon');
    return RoutineCompletion.fromJson(response.data as Map<String, dynamic>);
  }

  // ==================== Analytics ====================

  Future<RoutineAnalytics> getAnalytics({
    required int routineId,
    required String startDate,
    required String endDate,
  }) async {
    final response = await _apiService.get(
      '/routines/$routineId/analytics?startDate=$startDate&endDate=$endDate',
    );
    return RoutineAnalytics.fromJson(response.data as Map<String, dynamic>);
  }

  Future<List<RoutineHistory>> getHistory({
    required int routineId,
    required String startDate,
    required String endDate,
  }) async {
    final response = await _apiService.get(
      '/routines/$routineId/history?startDate=$startDate&endDate=$endDate',
    );
    final List<dynamic> data = response.data as List<dynamic>;
    return data.map((json) => RoutineHistory.fromJson(json as Map<String, dynamic>)).toList();
  }
}

// Singleton instance
final routineApi = RoutineApi(apiService);

File: web/src/services/routineApi.ts
import api from './api';
import type {
  Routine,
  RoutineDetail,
  RoutineStep,
  RoutineSchedule,
  RoutineCompletion,
  RoutineStepCompletion,
  RoutineAnalytics,
  RoutineHistory,
  PendingRoutinePrompt,
  CreateRoutineRequest,
  UpdateRoutineNameRequest,
  CreateRoutineStepRequest,
  UpdateRoutineStepTextRequest,
  UpdateRoutineStepNotesRequest,
  UpdateRoutineStepPositionRequest,
  SetRoutineSchedulesRequest,
  CompleteRoutineStepRequest,
} from '../types';

export const routineApi = {
  // ==================== Routine CRUD ====================

  async getAllRoutines(): Promise<Routine[]> {
    const response = await api.get<Routine[]>('/routines');
    return response.data;
  },

  async getRoutineDetail(routineId: number): Promise<RoutineDetail> {
    const response = await api.get<RoutineDetail>(`/routines/${routineId}`);
    return response.data;
  },

  async createRoutine(name: string): Promise<Routine> {
    const request: CreateRoutineRequest = { name };
    const response = await api.post<Routine>('/routines', request);
    return response.data;
  },

  async updateRoutineName(routineId: number, name: string): Promise<Routine> {
    const request: UpdateRoutineNameRequest = { name };
    const response = await api.put<Routine>(`/routines/${routineId}/name`, request);
    return response.data;
  },

  async deleteRoutine(routineId: number): Promise<{ message: string }> {
    const response = await api.delete<{ message: string }>(`/routines/${routineId}`);
    return response.data;
  },

  // ==================== Step CRUD ====================

  async createStep(routineId: number, text: string, notes?: string, position?: number): Promise<RoutineStep> {
    const request: CreateRoutineStepRequest = { text, notes, position };
    const response = await api.post<RoutineStep>(`/routines/${routineId}/steps`, request);
    return response.data;
  },

  async updateStepText(routineId: number, stepId: number, text: string): Promise<RoutineStep> {
    const request: UpdateRoutineStepTextRequest = { text };
    const response = await api.put<RoutineStep>(`/routines/${routineId}/steps/${stepId}/text`, request);
    return response.data;
  },

  async updateStepNotes(routineId: number, stepId: number, notes: string | null): Promise<RoutineStep> {
    const request: UpdateRoutineStepNotesRequest = { notes };
    const response = await api.put<RoutineStep>(`/routines/${routineId}/steps/${stepId}/notes`, request);
    return response.data;
  },

  async updateStepPosition(routineId: number, stepId: number, position: number): Promise<RoutineStep> {
    const request: UpdateRoutineStepPositionRequest = { position };
    const response = await api.put<RoutineStep>(`/routines/${routineId}/steps/${stepId}/position`, request);
    return response.data;
  },

  async deleteStep(routineId: number, stepId: number): Promise<{ message: string }> {
    const response = await api.delete<{ message: string }>(`/routines/${routineId}/steps/${stepId}`);
    return response.data;
  },

  // ==================== Schedules ====================

  async setSchedules(routineId: number, schedules: SetRoutineSchedulesRequest['schedules']): Promise<RoutineSchedule[]> {
    const request: SetRoutineSchedulesRequest = { schedules };
    const response = await api.put<RoutineSchedule[]>(`/routines/${routineId}/schedules`, request);
    return response.data;
  },

  // ==================== Prompts ====================

  async getPendingPrompts(): Promise<PendingRoutinePrompt[]> {
    const response = await api.get<PendingRoutinePrompt[]>('/routines/prompts/pending');
    return response.data;
  },

  async dismissPrompt(routineId: number): Promise<{ message: string }> {
    const response = await api.post<{ message: string }>(`/routines/prompts/${routineId}/dismiss`);
    return response.data;
  },

  // ==================== Execution ====================

  async quickCompleteRoutine(routineId: number, completedStepIds?: number[]): Promise<RoutineCompletion> {
    const body = completedStepIds ? { completedStepIds } : null;
    const response = await api.post<RoutineCompletion>(`/routines/${routineId}/quick-complete`, body);
    return response.data;
  },

  async startRoutine(routineId: number): Promise<RoutineCompletion> {
    const response = await api.post<RoutineCompletion>(`/routines/${routineId}/start`);
    return response.data;
  },

  async getActiveExecution(routineId: number): Promise<RoutineCompletion | null> {
    const response = await api.get<RoutineCompletion>(`/routines/${routineId}/active`);
    // Returns 204 No Content if no active execution
    return response.status === 204 ? null : response.data;
  },

  async completeStep(
    completionId: number,
    stepId: number,
    action: 'complete' | 'skip',
    notes?: string
  ): Promise<RoutineStepCompletion> {
    const request: CompleteRoutineStepRequest = { action, notes };
    const response = await api.post<RoutineStepCompletion>(
      `/routines/executions/${completionId}/steps/${stepId}`,
      request
    );
    return response.data;
  },

  async finishExecution(completionId: number): Promise<RoutineCompletion> {
    const response = await api.post<RoutineCompletion>(`/routines/executions/${completionId}/finish`);
    return response.data;
  },

  async abandonExecution(completionId: number): Promise<RoutineCompletion> {
    const response = await api.post<RoutineCompletion>(`/routines/executions/${completionId}/abandon`);
    return response.data;
  },

  // ==================== Analytics ====================

  async getAnalytics(routineId: number, startDate: string, endDate: string): Promise<RoutineAnalytics> {
    const response = await api.get<RoutineAnalytics>(
      `/routines/${routineId}/analytics?startDate=${startDate}&endDate=${endDate}`
    );
    return response.data;
  },

  async getHistory(routineId: number, startDate: string, endDate: string): Promise<RoutineHistory[]> {
    const response = await api.get<RoutineHistory[]>(
      `/routines/${routineId}/history?startDate=${startDate}&endDate=${endDate}`
    );
    return response.data;
  },
};
